core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1031 @@
+		EMSG("Invalid alg_type[%"PRIu32"]", msg->alg_type);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+}
+
+static enum hisi_drv_status hpre_rsa_fill_sqe(void *bd, void *info)
+{
+	struct hpre_rsa_msg *msg = info;
+	struct hpre_sqe *sqe = bd;
+
+	sqe->w0 = msg->alg_type | SHIFT_U32(0x1, HPRE_DONE_SHIFT);
+	sqe->task_len1 = TASK_LENGTH(msg->key_bytes);
+
+	return hpre_rsa_fill_addr_params(msg, sqe);
+}
+
+static enum hisi_drv_status hpre_rsa_parse_sqe(void *bd, void *info __unused)
+{
+	struct hpre_sqe *sqe = bd;
+	uint16_t err = 0;
+	uint16_t done = 0;
+
+	err = HPRE_TASK_ETYPE(sqe->w0);
+	done = HPRE_TASK_DONE(sqe->w0);
+	if (done != HPRE_HW_TASK_DONE || err) {
+		EMSG("HPRE do rsa fail! done=0x%"PRIX16", etype=0x%"PRIX16,
+		     done, err);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static TEE_Result hpre_rsa_do_task(void *msg)
+{
+	struct hisi_qp *rsa_qp = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	rsa_qp = hpre_create_qp(HISI_QM_CHANNEL_TYPE0);
+	if (!rsa_qp) {
+		EMSG("Fail to create rsa qp");
+		return TEE_ERROR_BUSY;
+	}
+
+	rsa_qp->fill_sqe = hpre_rsa_fill_sqe;
+	rsa_qp->parse_sqe = hpre_rsa_parse_sqe;
+	ret = hisi_qp_send(rsa_qp, msg);
+	if (ret) {
+		EMSG("Fail to send task, ret=%x"PRIx32, ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+	ret = hisi_qp_recv_sync(rsa_qp, msg);
+	if (ret) {
+		EMSG("Recv task error, ret=%x"PRIx32, ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+done_proc:
+	hisi_qm_release_qp(rsa_qp);
+
+	return res;
+}
+
+static TEE_Result mgf_process(size_t digest_size, uint8_t *seed,
+			      size_t seed_len, uint8_t *mask, size_t mask_len,
+			      struct drvcrypt_rsa_ed *rsa_data)
+{
+	struct drvcrypt_rsa_mgf mgf = { };
+
+	if (!rsa_data->mgf) {
+		EMSG("mgf function is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mgf.hash_algo = rsa_data->hash_algo;
+	mgf.digest_size = digest_size;
+	mgf.seed.data = seed;
+	mgf.seed.length = seed_len;
+	mgf.mask.data = mask;
+	mgf.mask.length = mask_len;
+
+	return rsa_data->mgf(&mgf);
+}
+
+static TEE_Result xor_process(uint8_t *a, uint8_t *b, uint8_t *out, size_t len)
+{
+	struct drvcrypt_mod_op xor_mod = { };
+
+	xor_mod.n.length = len;
+	xor_mod.a.data = a;
+	xor_mod.a.length = len;
+	xor_mod.b.data = b;
+	xor_mod.b.length = len;
+	xor_mod.result.data = out;
+	xor_mod.result.length = len;
+
+	return drvcrypt_xor_mod_n(&xor_mod);
+}
+
+static size_t hpre_rsa_get_hw_kbytes(size_t key_bits)
+{
+	size_t size = 0;
+
+	if (key_bits <= 1024)
+		size = BITS_TO_BYTES(1024);
+	else if (key_bits <= 2048)
+		size = BITS_TO_BYTES(2048);
+	else if (key_bits <= 3072)
+		size = BITS_TO_BYTES(3072);
+	else if (key_bits <= 4096)
+		size = BITS_TO_BYTES(4096);
+	else
+		EMSG("Invalid key_bits[%zu]", key_bits);
+
+	return size;
+}
+
+static void hpre_rsa_params_free(struct hpre_rsa_msg *msg)
+{
+	switch (msg->alg_type) {
+	case HPRE_ALG_KG_CRT:
+		free_wipe(msg->in);
```

Comment:
I don't see where `msg->in` is an allocated pointer. I think at this stage, `msg->in` is always NULL.

Reply:
solved